### DISCUSSION TITLE
Berry tasmota.add_cmd in a class

### DISCUSSION THREAD

**Discussion by DetlefR** [arendst/Tasmota#16450] [Q&A] [answered]
Hello,

I try to implement a cmnd inside a berry class.
What I have do:
I have created a class
```
class MYCLASS
 def init()
 end
  
 def cmnd(cmd, idx, payload, payload_json)
  print(cmd, idx, payload, payload_json)
  tasmota.resp_cmnd_done()
 end
end
```
make an instance and register the command
```
var myclass=MYCLASS()
tasmota.add_cmd('mycmd',/-> myclass.cmnd())
```
A test from the console ```mycmd {"Hallo":123} ```shows me, the cmnd function is called. The RESULT is {"mycmd":"Done"). But print brings four times "nil"
How must a cmd correct registered?

**Comment by s-hadinger** [comment, marked as answer]
You need to capture the arguments when you create the closure:

``` berry
var myclass=MYCLASS()
tasmota.add_cmd('mycmd',/ cmd, idx, payload, payload_json -> myclass.cmnd(cmd, idx, payload, payload_json))
```

**Reply by DetlefR** [reply]
@s-hadinger 
Thanks. This is working.

**Reply by ChrisVonFrosta** [reply]
or do `cmnd` static?!
```
class MYCLASS
 def init()
 end
  
static def cmnd(cmd, idx, payload, payload_json)
  print(cmd, idx, payload, payload_json)
  tasmota.resp_cmnd_done()
 end
end
```

**Reply by sfromis** [reply]
Not really recommended, even if it *can* work in isolation. Making it static defeats a main purpose of creating a class, to have a class instance with status in variables, with the method thereby participating in handling of status data, like if you want to use the command to update something the class is using, or have the command method take action based on current status.

**Comment by ChrisVonFrosta** [comment]
I use it this way to receive parameters as a list of maps and execute some methods from another class using these parameters. But I understand your point: it needs to be non-static if you require a reference to self.

However, what I do wonder about is if, when calling a command via an HTTP request, all print outputs appear in the response data until tasmota.resp_cmnd_string(response_json_string) is executed.

**Reply by sfromis** [reply]
Not clear what you specifically mean by that.

In general you should always aim to make command processing very fast, including generating the response, as the whole of Tasmota is blocked while the callback code is running. That also means that you should **not** do things like call `tasmota.cmd` from inside the command processing callback, as this is likely to cause breakage.

**Comment by ChrisVonFrosta** [comment]
something like that:

```
def test(cmd, idx, payload, payload_json)
  prin(payload_json)
  tasmota.resp_cmnd_str('{"response":"ok"}')
def

tasmota.add_cmd("test",test)
```

calling via webbrowser:
`htttp:<tasmota_ip>/cm?cmnd=test{"hallo":1}`

will response something like that:
`{"hallo":1, "test": {"response": "ok"}}`


what i mean is: i got some of my debug prints in the result response, if calling a command remotely via http request

**Reply by sfromis** [reply]
Except for the typos in what you posted there, you get all of the command response. What you supply with the call from the command function is the *result* message from command execution, not the whole command *response*, which can indeed be also multiple lines from some existing Tasmota commands.

Anyway, when you want to have the command response as JSON, the intention is to use `tasmota.resp_cmnd` instead of `tasmota.resp_cmnd_str`, to simply have the result be what you constructed.

**Reply by ChrisVonFrosta** [reply]
I'm not quite following and probably didn't express myself clearly.
I had assumed that the response for the other party (client) only comes from tasmota.resp_cmd() or tasmota.resp_cmd_str(). However, I receive all print statements in the response. If every previous debug output of these data is included in the response, it makes debugging more cumbersome.

**Reply by ChrisVonFrosta** [reply]
i have the following:

`tasmota.add_cmd("display_http_request", Display.http_request)`

with:
```
class Display

  static def http_request(cmd, idx, payload, payload_json)
      ...
      return_list = Aktor.list_all(aktortype, room, id)   # returns list of maps
      # print(f"return_list  {return_list}")               # client would see this. WHY?
      # tasmota.resp_cmnd_str(return_list)           # this works, if client do parsing and converting correctly

      var return_list_json = json.dump(return_list)      # this results in a "perfect" json
      # print(f"return_list_json  {return_list_json}")      # client would see this. WHY?

      tasmota.resp_cmnd(return_list_json)              #i can't "parse" this on client side nativly, client side lose the `[]` of the list of maps
   end
end
```
console output on http request:
```
Grp 0, Cmd 'DISPLAY_HTTP_REQUEST', Idx 1, Len 43, Pld -99, Data '{"request_AktorList":{"aktortype":"Licht"}}'
01:20:47.037 BRY: GC from 40836 to 31008 bytes, objects freed 231/410 (in 2 ms) - slots from 761/774 to 388/609
01:20:47.042 RSL: RESULT = [{"room":"Träger","output_number":2,"name":"Decko","floor":"1","aktortype":"Licht","id":9},{"room":"TrägerAB","output_number":3,"name":"DeckoAB","floor":"1","aktortype":"Licht","id":10}]
```

on client side:
```
static def serverrequest()
  ...
  
  print(f"url_string   {url_string}")
    
  cl.begin(url_string)
  
  var r = cl.GET()
  print(r)

  var response_string = cl.get_string()                          #loosing some json specs?
  print(f"response_string   {response_string}")      

  var response_json = json.load(response_string)              # becomes nil
#  var response_json = json.load("["+ response_string + "]")   # now i got it work this way, the return is a needed list of maps 
  print(f"response_json   {response_json}")         

  return response_json 
end
```


 output:
```
url_string   http://192.168.170.68/cm?cmnd=display_http_request{"request_AktorList":{"aktortype":"Licht"}}
200
response_string   {"room":"Träger","output_number":2,"name":"Decko","floor":"1","aktortype":"Licht","id":9},{"room":"TrägerAB","output_number":3,"name":"DeckoAB","floor":"1","aktortype":"Licht","id":10}
response_json   nil
```

The client loses the brackets of the list (or the array, in JSON terminology).
Is there allways the need to reconstruct the received json?

**Reply by sfromis** [reply]
On a more architectural level, the issue seems to be that you want to see the command processing as a way to create an API. It is not. It is meant for "human-level" commands with simple arguments and responses, not to transfer complex objects not meant for direct human consumption.

The Tasmota mechanisms for handling commands/responses are structured around how the existing JSON handling is done in Tasmota, not for processing arbitrary JSON like arrays at top level. That also means that Tasmota is expecting the JSON responses provided being suitable for processing by Tasmota "classic" rules, which yours are not. If you run with `Weblog 3` enabled, you should see error messages about "No valid JSON" in the console, indicating that you're stepping outside of what Tasmota is meant to handle.

When you do not want Tasmota JSON handling of your JSON, you could revert to `tasmota.resp_cmnd_str`, meaning that your JSON gets wrapped in a string inside JSON with the command name as key, which you then could unpick. While this is not great for "regular" command responses, you do get the raw string....

When it comes to the "archictural" level, more suitable Berry mechanisms for transferring data is "anything but commands", but instead MQTT, TCP, UDP or webserver/webclient to pass "raw" data.

And it is not a bug, but a feature with with multiline responses being gathered into one message, including your result data. This is not specific to Berry, but a more general feature. It might have surprised you, but that's life ...

**Comment by ChrisVonFrosta** [comment]
i have to work through your explanations but i want to say directly, thank you for spending time to response so extensive!

i was thinking about switiching to tcp or webserver/webclient. But since i started the "proof of conecpt" with this low-entry-level 'tasmota.add.cmd()` option, i extended this for next commands.

You're right. I ultimately expanded it into an API because I didn't want to add a new command for every additional request (tasmota.add.cmd()). So, you recommend switching to one of the methods you mentioned. However, I can't natively transfer objects or lists with those methods, can I? They always involve transmitting bytes/strings, and I have to parse them myself, right?

**Reply by sfromis** [reply]
You can easily transform maps or lists to strings, and "less simple" objects can have a `tostring()` method to transform to a string representation (preferably JSON), which then would have to be parsed on the other end. If you structure your context as a map with fields, this is easy to serialize and restore. Of course, things gets more complex if/when you have objects with references to other objects, as you can only get simple strings/numbers/booleans/lists/maps represented as JSON.

Sure, you *could* invent some binary format, but that is rarely worth it, with JSON well-supported and reasonably readable too.

**Reply by ChrisVonFrosta** [reply]
can i send MQTT data between 2 Tasmota device without a broker?
What would be the fastes way of inter-tasmota-communication?
what would your favour for such data?

**Reply by sfromis** [reply]
MQTT always needs a broker, and is a good communication strategy. Probably even faster than cobbling something with http, which needs extra roundtrips for every session. MQTT is also automatically asynchronous, meaning that the sender will not be slowed down by the process.

Using UDP can be faster, but you if you actually want reliable communication, you need to create a protocol to ensure retransmission, to handle lost data, meaning that the speed advantage is not so much after all. And no, do not think that lost data is no issue on a local Wifi network.

**Reply by ChrisVonFrosta** [reply]
Thank you.
Since I prefer not to use a separate broker and do not want to manage retransmissions, I would opt for TCP or WebClient. How do they compare in terms of speed?

**Reply by sfromis** [reply]
Internally, simple http is nothing but tcp, just with some conventions about message structure. That means that it is basically as slow, due to having to wait for the other end to respond. That's why it is generally slower than MQTT. However, with TCP you *can* create a protocol which keeps a session running, instead of a new one every time.

Still, you have https://tasmota.github.io/docs/Berry/#tcpclientasync-class - of course with similar protocol considerations as the inherently unreliable UDP.

**Reply by ChrisVonFrosta** [reply]
thank you again and again and...!
